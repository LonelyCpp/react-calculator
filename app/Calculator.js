export default class Calculator {
  static getPriority = char => {
    switch (char) {
      case '+':
      case '-':
        return 1;
      case '*':
      case '/':
        return 2;
      default:
        return 0;
    }
  };

  static getStackTop = stack => {
    if (stack.length > 0) {
      return stack[stack.length - 1];
    }
  };

  static infixToPostfix = infix => {
    infix = infix.replace(/\+/g, ' + ');
    infix = infix.replace(/\-/g, ' - ');
    infix = infix.replace(/\//g, ' / ');
    infix = infix.replace(/\*/g, ' * ');

    infix = infix.split(' ');

    let opStack = [];
    let postfix = [];

    infix.forEach(element => {
      switch (element) {
        case '+':
        case '-':
        case '*':
        case '/': {
          if (opStack.length == 0) {
            opStack.push(element);
          } else {
            let stTopPriority = Calculator.getPriority(
              Calculator.getStackTop(opStack)
            );
            let elePriority = Calculator.getPriority(element);

            if (elePriority >= stTopPriority) {
              opStack.push(element);
            } else {
              while (elePriority < stTopPriority && opStack.length != 0) {
                postfix.push(opStack.pop());
                stTopPriority = Calculator.getPriority(
                  Calculator.getStackTop(opStack)
                );
              }
              opStack.push(element);
            }
          }
          break;
        }
        default:
          postfix.push(parseFloat(element));
      }
    });

    while (opStack.length != 0) {
      postfix.push(opStack.pop());
    }

    return postfix;
  };

  static evalPostfix(postfix) {
    let evalStack = [];

    postfix.forEach(element => {
      try {
        switch (element) {
          case '+': {
            let v1 = evalStack.pop();
            let v2 = evalStack.pop();
            evalStack.push(v2 + v1);
            break;
          }
          case '-': {
            let v1 = evalStack.pop();
            let v2 = evalStack.pop();
            evalStack.push(v2 - v1);
            break;
          }
          case '*': {
            let v1 = evalStack.pop();
            let v2 = evalStack.pop();
            evalStack.push(v2 * v1);
            break;
          }
          case '/': {
            let v1 = evalStack.pop();
            let v2 = evalStack.pop();
            evalStack.push(v2 / v1);
            break;
          }
          default: {
            evalStack.push(element);
          }
        }
      } catch (err) {
        console.error(err);
      }
    });

    return Calculator.getStackTop(evalStack);
  }
}
