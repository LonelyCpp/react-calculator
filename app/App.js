// @flow
import React, { Component } from 'react';
import { Animated, Text, View } from 'react-native';

import mainLayoutStyle from './styles/mainLayout.js';
import textStyle from './styles/text.js';
import NumberButton from './NumberButton.js';

import Calculator from './Calculator.js';
import colors from './styles/colors.js';

export default class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      expression: '',
      clear: false
    };
  }

  isOperator = char => {
    switch (char) {
      case '*':
      case '+':
      case '/':
      case '-':
        return true;
    }
    return false;
  };

  _onPress = char => {
    console.log(char);
    // Clears the notepad if it contains an answer
    if (this.state.clear) {
      this.state.expression = '';
      this.state.clear = false;
    }
    // Blank button press
    if (char == '') {
      return;
    }
    // Calculate answer
    else if (char == '=') {
      let postfix = Calculator.infixToPostfix(this.state.expression);
      let ans = Calculator.evalPostfix(postfix);
      if (isNaN(ans)) {
        ans = 'Invalid Expression';
      }
      this.setState({ expression: ans });
      this.state.clear = true;
    }
    // Clear notepad
    else if (char == 'C') {
      this.setState(previousState => ({ expression: '' }));
    }
    // Backspace
    else if (char == '<') {
      this.setState(previousState => ({
        expression: previousState.expression.slice(0, -1)
      }));
    }
    // operators or numbers
    else {
      let len = this.state.expression.length;
      if (this.isOperator(char)) {
        // checks if the last character was an operator
        // avoids appending two consecutive operators
        if (len > 0) {
          let lastChar = this.state.expression[len - 1];
          if (this.isOperator(lastChar)) {
            this.setState(previousState => ({
              expression: previousState.expression.slice(0, -1)
            }));
          }
        } else {
          return;
        }
      }
      this.setState(previousState => ({
        expression: previousState.expression + char
      }));
    }
  };

  render() {
    return (
      <View style={mainLayoutStyle.container}>
        <View style={mainLayoutStyle.notepadRow}>
          <Text style={textStyle.notepad}>{this.state.expression}</Text>
        </View>
        <View style={mainLayoutStyle.numberpadRow}>
          <NumberButton
            text="C"
            onPress={this._onPress}
            textColor={colors.THEME_LIGHT_SECONDARY}
          />
          <NumberButton text="" onPress={this._onPress} />
          <NumberButton
            text="<"
            onPress={this._onPress}
            textColor={colors.THEME_LIGHT_SECONDARY}
          />
          <NumberButton
            text="/"
            onPress={this._onPress}
            textColor={colors.THEME_LIGHT_SECONDARY}
          />
        </View>
        <View style={mainLayoutStyle.numberpadRow}>
          <NumberButton text="7" onPress={this._onPress} />
          <NumberButton text="8" onPress={this._onPress} />
          <NumberButton text="9" onPress={this._onPress} />
          <NumberButton
            text="*"
            onPress={this._onPress}
            textColor={colors.THEME_LIGHT_SECONDARY}
          />
        </View>
        <View style={mainLayoutStyle.numberpadRow}>
          <NumberButton text="4" onPress={this._onPress} />
          <NumberButton text="5" onPress={this._onPress} />
          <NumberButton text="6" onPress={this._onPress} />
          <NumberButton
            text="-"
            onPress={this._onPress}
            textColor={colors.THEME_LIGHT_SECONDARY}
          />
        </View>
        <View style={mainLayoutStyle.numberpadRow}>
          <NumberButton text="1" onPress={this._onPress} />
          <NumberButton text="2" onPress={this._onPress} />
          <NumberButton text="3" onPress={this._onPress} />
          <NumberButton
            text="+"
            onPress={this._onPress}
            textColor={colors.THEME_LIGHT_SECONDARY}
          />
        </View>
        <View style={mainLayoutStyle.numberpadRow}>
          <NumberButton text="" onPress={this._onPress} />
          <NumberButton text="0" onPress={this._onPress} />
          <NumberButton text="." onPress={this._onPress} />
          <NumberButton
            text="="
            onPress={this._onPress}
            textColor={colors.THEME_LIGHT_SECONDARY}
          />
        </View>
      </View>
    );
  }
}
