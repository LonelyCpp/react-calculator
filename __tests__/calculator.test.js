import React from 'react';
import Calculator from '../app/Calculator';
import App from '../app/App'

import renderer from 'react-test-renderer';

test('infix to postfix', () => {
    expect(Calculator.infixToPostfix("9+9*9+9")).toEqual([9, 9, 9, "*", 9, "+", "+"]);
    expect(Calculator.infixToPostfix("9/9+9*9")).toEqual([9, 9, "/", 9, 9, "*", "+"]);
    expect(Calculator.infixToPostfix("9+9*9")).toEqual([9, 9, 9, "*", "+"]);
    expect(Calculator.infixToPostfix("9*9-9")).toEqual([9, 9, "*", 9, "-"]);    
});

test('postfix evaluation', () => {
  expect(Calculator.evalPostfix([9, 9, '+'])).toBe(18);
  expect(Calculator.evalPostfix([9, 1, '+'])).toBe(10);
  expect(Calculator.evalPostfix([3, 2, 5, '*', '+'])).toBe(13);
  expect(Calculator.evalPostfix([9, '+'])).toBe(NaN);
});

test('renders correctly', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree).toMatchSnapshot();
  });