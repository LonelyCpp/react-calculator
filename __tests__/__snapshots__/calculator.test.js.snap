// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#F5FCFF",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-end",
        "alignSelf": "stretch",
        "backgroundColor": "orange",
        "borderBottomWidth": 3,
        "borderColor": "#000000",
        "flex": 3,
        "justifyContent": "flex-end",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 50,
        }
      }
    >
      
    </Text>
  </View>
  <View
    style={
      Object {
        "alignSelf": "stretch",
        "backgroundColor": "white",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignSelf": "stretch",
        "backgroundColor": "white",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignSelf": "stretch",
        "backgroundColor": "white",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignSelf": "stretch",
        "backgroundColor": "white",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignSelf": "stretch",
        "backgroundColor": "white",
        "flex": 1,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffbcbc",
            "borderColor": "red",
            "borderWidth": 1,
            "height": 100,
            "justifyContent": "center",
            "margin": 10,
            "width": 300,
          },
          undefined,
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#333333",
            "margin": 20,
          }
        }
      >
        TouchableNativeFeedback is not supported on this platform!
      </Text>
    </View>
  </View>
</View>
`;
